### PR TITLE
ext_proc: Support per-route GrpcService configuration.

### PR DESCRIPTION
**Pull request by mpwarres**
This adds the ability to change the GrpcService used by the ext_proc filter on a per-route basis.

Signed-off-by: Michael Warres <mpw@google.com>

Commit Message: ext_proc: Support per-route GrpcService configuration
Additional Description:
Risk Level: Low. Not triggered unless configured.
Testing: New unit and integration tests added.
Docs Changes: Addition of new config field.
Release Notes:
Platform Specific Features:

### CODE REVIEW

**Comment by mpwarres** [comment]
Requesting review from ext_proc owners.

/assign @chaoqin-li1123 @pradeepcrao

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 1]
```diff
@@ -705,19 +711,26 @@ void Filter::mergePerRouteConfig() {
...
+    return;
+  }
+  if (merged_config->processingMode()) {
+    ENVOY_LOG(trace, "Setting new processing mode from per-route configuration");
+    decoding_state_.setProcessingMode(*(merged_config->processingMode()));
+    encoding_state_.setProcessingMode(*(merged_config->processingMode()));
+  }
+  if (merged_config->grpcService()) {
+    ENVOY_LOG(trace, "Setting new GrpcService from per-route configuration");
+    grpc_service_ = *merged_config->grpcService();
```
Comment (htuch) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:733]: Can the route override change the gRPC service after the initial connection?

**Review by htuch — COMMENTED** [review]
How much scalability in # services do you need? I don't think this is going to be any different than regular Envoy cluster or gRPC client scalability concerns, but worth keeping in mind if you have many different services across routes.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 1 above)
Comment (mpwarres) [reply source/extensions/filters/http/ext_proc/ext_proc.cc:733]: No--the mechanism proposed here is that the ext_proc filter uses a single grpc_service for the entirety of a given HTTP request/response. This is an extension of how ext_proc currently handles other per-route config, which is also latched once, at decodeHeaders.

**Comment by mpwarres** [comment]
> How much scalability in # services do you need? I don't think this is going to be any different than regular Envoy cluster or gRPC client scalability concerns, but worth keeping in mind if you have many different services across routes.

As discussed offline, we don't anticipate a large number of services in the near term, and will look to other approaches to address future scaling, such as having ExtProc set metadata based on per-route config that could then be read and acted upon by a (single) grpc_service's upstream filter.

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto [hunk 2]
```diff
@@ -195,4 +196,7 @@ message ExtProcOverrides {
   // Set different optional properties than the default setting of the
   // ``response_attributes`` field.
   repeated string response_attributes = 4;
+
+  // Set a different gRPC service for this route than the default.
+  config.core.v3.GrpcService grpc_service = 5;
```
Comment (htuch) [inline api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:201]: Please add to version history.

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc [hunk 3]
```diff
@@ -291,12 +294,11 @@ class ExtProcIntegrationTest : public HttpIntegrationTest,
     }
   }
 
-  void processAndRespondImmediately(bool first_message,
+  void processAndRespondImmediately(FakeUpstream& grpc_upstream, bool first_message,
```
Comment (htuch) [inline test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:297]: Nit: do you think organizing these gRPC upstream methods inside a class would make sense?

**Review by htuch — COMMENTED** [review]
LGTM. @snowp can you take a non-Googler pass over this for general applicability? Thanks.

File: source/extensions/filters/http/ext_proc/ext_proc.cc [hunk 4]
```diff
@@ -61,11 +61,17 @@ FilterConfigPerRoute::FilterConfigPerRoute(const ExtProcPerRoute& config)
...
   }
+  if (config.overrides().has_grpc_service()) {
+    grpc_service_ = config.overrides().grpc_service();
+  }
 }
 
 void FilterConfigPerRoute::merge(const FilterConfigPerRoute& src) {
   disabled_ = src.disabled_;
   processing_mode_ = src.processing_mode_;
+  if (src.grpcService()) {
```
Comment (chaoqin-li1123) [inline source/extensions/filters/http/ext_proc/ext_proc.cc:72]: Call has_value() to be explicit.

**Review by chaoqin-li1123 — DISMISSED** [review]

File: api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto (hunk 2 above)
Comment (mpwarres) [reply api/envoy/extensions/filters/http/ext_proc/v3/ext_proc.proto:201]: Done.

File: source/extensions/filters/http/ext_proc/ext_proc.cc (hunk 4 above)
Comment (mpwarres) [reply source/extensions/filters/http/ext_proc/ext_proc.cc]: Done.

File: test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc (hunk 3 above)
Comment (mpwarres) [reply test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:297]: Yes, I think that would be cleaner, and allow for more flexible tests in the future with separate processor_connection_ and processor_stream_ for each FakeExternalProcessor.

I'd prefer to do that in a follow-on PR, though, to delineate the test refactor from the behavior additions in this PR.

Comment (htuch) [reply test/extensions/filters/http/ext_proc/ext_proc_integration_test.cc:297]: Sure, that's fine.

**Comment by htuch** [comment]
@snowp PTAL when you get a chance, thanks.

**Review by snowp — APPROVED** [review]
Makes sense to me, thanks!

**Comment by snowp** [comment]
Needs main merge

/wait

**Comment by mpwarres** [comment]
Thanks for the reviews! Have merged from main.

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Pull request merged**